### ISSUE TITLE
@protected should allow same-file access

### ISSUE THREAD

**Issue by Hixie** [dart-lang/linter#254]
Right now @protected is actually stricter than _private, in that it even disallows same-file access. I'd be totally ok with @strictProtected and @strictPrivate modifiers that prevent access from outside the class entirely, but I think the @protected annotation should default to being a superset of _private, i.e. only additionally allowing access from outside the file to subclasses.

**Comment by pq** [comment]
By outside the file I think you mean "outside the library" (e.g., what about parts)?  Are you looking for something more like Java's notion of `protected`?

**Comment by zoechi** [comment]
I also think `@protected` should not be more limiting than private anywhere. It should only restrict public acces to superclasses. I now get a warning because I try to access a protected field in the same class :D

**Comment by bwilkerson** [comment]
> ...  I think the `@protected` annotation should default to being a superset of _private ...

That might have been the motivation behind Java's decision to define `protected` that way, but in practice I never saw anyone use it to mean anything other than "visible to subclasses of this class". Do you have a use case in mind for wanting to make these members visible to other classes in the same library?

> I now get a warning because I try to access a protected field in the same class.

Well, that's clearly a bug in the implementation of the rule.

**Comment by pq** [comment]
> I now get a warning because I try to access a protected field in the same class 

@zoechi is this with the latest dev build or HEAD?  I *think* have a test for this but to be sure maybe you could share a repro?

As always, thanks!

**Comment by zoechi** [comment]
@pq

Dart VM version: 1.17.0-edge.6b6d363900373a56caa971b5610b3a0050ed55f6 (Thu Jun  2 17:30:30 2016) on "linux_x64"

```
library bwu_datagrid.plugins;

import 'package:meta/meta.dart' show protected;
import 'package:bwu_datagrid/bwu_datagrid.dart';
import 'package:bwu_datagrid/core/core.dart' as core;

abstract class Plugin {
  BwuDatagrid _grid;
  BwuDatagrid get grid => _grid;
  @protected
  set grid(BwuDatagrid value) => _grid = value;
  bool isSuspended = false;

  Plugin();

  void init(BwuDatagrid grid) {
    this.grid = grid; // <<<=== error for protected access
  }
  void destroy() {}
}

abstract class SelectionModel extends Plugin {
  SelectionModel();

  void setSelectedRanges(List<core.Range> ranges);
  List<core.Range> getSelectedRanges();
}
```

**Comment by pq** [comment]
👍  Thanks for the repro @zoechi !

Working on a fix --- tracking here: https://github.com/dart-lang/sdk/issues/26614.

**Comment by zoechi** [comment]
@bwilkerson 

> use case in mind for wanting to make these members visible to other classes in the same library?

I currently don't have a concrete use case in mind, just found it weird, maybe just of different expectations. I'll post an update if I find one ...

**Comment by Hixie** [comment]
@pq I have no opinion on parts. I think they're a misfeature and we don't use them in flutter. 🙂

**Comment by Hixie** [comment]
You can see use cases for making things visible to other classes (and globals) in the same file in flutter's `framework.dart`. The internals of the widget framework is a close choreography of three class hierarchies, but we still need to have some stuff hidden from the users of these classes.

**Comment by pq** [comment]
via @Hixie, the [Object Pascal precedent](http://www.freepascal.org/docs-html/ref/refse33.html):

> The visibility of the different sections is as follows:
> 
> **Private** 
> All fields and methods that are in a private block, can only be accessed in the module (i.e. unit) that contains the class definition. They can be accessed from inside the classes’ methods or from outside them (e.g. from other classes’ methods)
> **Strict Private** 
> All fields and methods that are in a strict private block, can only be accessed from methods of the class itself. Other classes or descendent classes (even in the same unit) cannot access strict private members.
> **Protected** 
> Is the same as Private, except that the members of a Protected section are also accessible to descendent types, even if they are implemented in other modules.
> **Strict Protected** 
> Is the same as Protected, except that the members of a Protected section are also accessible to other classes implemented in the same unit. Strict protected members are only visible to descendent classes, not to other classes in the same unit.
> **Public** 
> sections are always accessible.
> **Published** 
> From a language perspective, this is the same as a Public section, but the compiler generates also type information that is needed for automatic streaming of these classes if the compiler is in the {$M+} state. Fields defined in a published section must be of class type. Array properties cannot be in a published section.

**Comment by pq** [comment]
Bumping priority to HIGH as the work-around (see: https://github.com/flutter/flutter/issues/4697) is a hack and degrades performance.

cc @bwilkerson

**Comment by alexeieleusis** [comment]
I would gladly volunteer to implement this, but I do not understand exactly what the problem is. If someone could provide a more concrete description of what and what is not needed I could start working on this once I am done with the current rule I am currently implementing.

**Comment by Hixie** [comment]
```dart
class A {
  @protected
  void m1() { }

  @strictProtected
  void m2() { }

  // private
  void _m3() { }

  @strictPrivate
  void _m4() { }

  // public
  void m5() { }
}

class B {
  void test(A a) {
    a.m1(); // this should not trigger an analyzer warning (this is the desired change)
    a.m2(); // this _should_ trigger an analyzer warning (for extra credit)
    a._m3(); // this should not trigger an analyzer warning (already implemented)
    a._m4(); // this _should_ trigger an analyzer warning (for extra credit)
    a.m5(); // this should not trigger an analyzer warning (already implemented)
  }
}
```

**Comment by Hixie** [comment]
Only the m1 case in the sample code above is the case we care about for Flutter. The others are just explaining how the FPC precedent @pq posted would map to Dart.

**Comment by Hixie** [comment]
Oh, and if the above is file1.dart, and the following is file2.dart:

```dart
import 'file1.dart';

class C {
  void test(A a) {
    a.m1(); // this _should_ trigger an analyzer warning (already implemented)
    a.m2(); // this _should_ trigger an analyzer warning (for extra credit)
    a._m3(); // calls noSuchMethod (already implemented)
    a._m4(); // calls noSuchMethod (already implemented)
    a.m5(); // this should not trigger an analyzer warning (already implemented)
  }
}
```

**Comment by alexeieleusis** [comment]
Thanks, I was close, but I think the change needs to be performed in the analyzer itself:
```
a.m1(); // this should not trigger an analyzer warning (this is the desired change)
```
I don't think I am familiar enough with that code base yet. I see this is labeled as a high priority, if you are OK with the time learning will take, I'll do it.

**Comment by pq** [comment]
> I would gladly volunteer to implement this

Thanks!  but you're all in luck, I've got this nearly ready for review. 😄

**Comment by cubuspl42** [comment]
I don't understand the rationale for this change. I acknowledge that "Flutter needed this", but I'm not sure if it's a good enough reason. The documentation for `@protected` explicitly states:

> Indicates that m should only be invoked from instance methods of C or classes that extend, implement or mix in C, either directly or indirectly. Additionally indicates that m should only be invoked on this, whether explicitly or implicitly.

Is says nothing about same library access. People expecting/needing the documented behaviour of `@protected` are now left without analyzer support.

**Comment by cubuspl42** [comment]
A potential fix (although not my favourite) would be to change the documentation of `@protected`.
